samples/remote_attestation/common/dispatcher.cpp

```diff
@@ -182,10 +182,173 @@ int ecall_dispatcher::verify_report_and_set_pubkey(
     return ret;
 }
 
+int ecall_dispatcher::establish_secure_channel(uint8_t** key, size_t* key_size)
+{
+    uint8_t encrypted_key_buf[512];
+    uint8_t signature[256];
+    uint8_t digest[32];
+    size_t encrypted_key_size = 256;
+    size_t signature_size = 256;
+    size_t total_size = 512;
+
+    int ret = 1;
+
+    if (m_initialized == false)
+    {
+        TRACE_ENCLAVE("ecall_dispatcher initialization failed.");
+        goto exit;
+    }
+
+    // Step 1 - Create an ephemeral symmetric key
+    if (oe_random(m_enclave_config->sym_key, 32) != OE_OK)
+    {
+        TRACE_ENCLAVE("ecall_dispatcher initialization failed to generate "
+                      "ephemeral symmetric key.");
+        goto exit;
+    }
+
+    TRACE_ENCLAVE("enclave: establish_secure_channel: Generated random "
+                  "symmetric key successfully");
+
+    // Step 2- Encrypt the symmetric key with the other enclave's public key
+    // TO DO - PREFIX with counter/seq number to prevent replay attacks
+    if (m_crypto->Encrypt(
+            m_crypto->get_the_other_enclave_public_key(),
+            m_enclave_config->sym_key,
+            32,
+            encrypted_key_buf,
+            &encrypted_key_size))
+
+    {
+        // Reason we allocate memory in the host is so that the encrypted data
+        // can be accessed by the hosts - aka real world scenario where we have
+        // 2 hosts and two enclaves
+        uint8_t* host_buf = (uint8_t*)oe_host_malloc(total_size);
+        if (host_buf == NULL)
+        {
+            TRACE_ENCLAVE(
+                "enclave: establish_secure_channel: oe_host_malloc failed.");
+            goto exit;
+        }
+
+        // Step 3 - Compute a SHA hash of the encrypted key
+        if (m_crypto->Sha256(encrypted_key_buf, encrypted_key_size, digest) !=
+            0)
+        {
+            goto exit;
+        }
+
+        TRACE_ENCLAVE("enclave: establish_secure_channel: Computed SHA hash "
+                      "of the encrypted key");
+
+        // Step 4 - Sign this SHA hash with my enclave's private key
+        if (m_crypto->Sign(digest, 32, signature, &signature_size) != 0)
```

Comment:
Doing it this way (rsa encrypt -> sign) means that the receiver cannot verify that the signed message actually came from the your enclave.

For example, if Alice, Bob and Charlie are trust each other and Alice sends an encrypted and signed message to Bob (i.e `Sign(Encrypt(Bob_pub, Message), Alice_private)`), then Charlie can intercept the message and replace the signature with his own (i.e `Sign(Encrypt, Bob_pub, Message), Charlie_private)`). They're no way for Alice to verify who actually sent the message without that info embedded in the message.

Reply:
You are right that Alice will not know who sent the original message if the signature gets switched by Charlie but the decryption of the message will fail as Charlie's private key will be used to decrypt. _One thing to note is that Charlie will need to first block Bob's receipt of Alice's original message, before sending the re-signed ciphertext._ 

There are issues with both methods of Sign & Encrypt vs Encrypt & Sign. 

So, best approach would be to use the Diffie-Hellman key exchange to establish the shared secret.

Reply:
@gupta-ak - You are right that Alice will not know who sent the original message. One thing to note is that Charlie will need to first block Bob's receipt of Alice's original message, before sending the re-signed ciphertext.

There are issues with both methods of Sign & Encrypt vs Encrypt & Sign.
Sign-Encrypt-Sign or Encrypt-Sign-Encrypt are suggested ways to resolve this.

So, best approach would be to use the Diffie-Hellman key exchange to establish the shared secret.

Reply:
@gupta-ak - Wanted to add that in this specific sample, Alice is only expecting message from Bob and will not proceed to decrypt the message if the signature is from some one else, including another trusted party. So, this suffices for this sample.

**Considering adding the name in the encrypted message so that in a general case, Alice can additionally verify that the signature and the encrypted was sent by the same trusted source.** 
Thanks.